### PR TITLE
SDPT-590 implement rest api calls in netty framework

### PR DESCRIPTION
**Pull request by vangork**
1. build up a http server based on netty framework
2. implement notetime call
3. implement write event call

### CODE REVIEW

File: src/main/java/io/pravega/ingest/gateway/server/http/HttpServer.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    public void start() throws Exception {
+        // Configure the server.
+        EventLoopGroup bossGroup = new NioEventLoopGroup(1);
+        EventLoopGroup workerGroup = new NioEventLoopGroup();
+        try {
+            ServerBootstrap b = new ServerBootstrap();
+            b.option(ChannelOption.SO_BACKLOG, 1024);
+            b.childOption(ChannelOption.TCP_NODELAY, true);
```
Comment (claudiofahey) [inline src/main/java/io/pravega/ingest/gateway/server/http/HttpServer.java:34]: Why are you setting this? If this is important, explain this in a code comment.

File: src/main/java/io/pravega/ingest/gateway/server/http/HttpServerRequest.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        // QueryStringDecoder queryDecoder = new QueryStringDecoder(jsonStr, false);
+        // Map<String, List<String>> uriAttributes = queryDecoder.parameters();
+    }
+
+    private byte[] decompress(byte[] in) throws Exception {
+        GZIPInputStream gzIn = new GZIPInputStream(new ByteArrayInputStream(in));
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        int res = 0;
+        byte buf[] = new byte[1024];
+        while (res >= 0) {
```
Comment (claudiofahey) [inline src/main/java/io/pravega/ingest/gateway/server/http/HttpServerRequest.java:108]: Consider using https://guava.dev/releases/19.0/api/docs/com/google/common/io/ByteStreams.html#copy(java.io.InputStream,%20java.io.OutputStream).

File: src/main/java/io/pravega/ingest/gateway/server/http/HttpServerRouter.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        }
+        else if(method.equals(HttpMethod.POST)) {
+            return revoke(path, postMap, request, response);
+        }
+        else {
+            return response.error(400, String.format("Http Method %s Not Supported Yet!", method.name()));
+        }
+    }
+
+    private HttpServerResponse revoke(String path, Map<String, RouteEntry> map, HttpServerRequest request, HttpServerResponse response) throws Exception {
```
Comment (claudiofahey) [inline src/main/java/io/pravega/ingest/gateway/server/http/HttpServerRouter.java:113]: Why is this called "revoke"? Maybe you mean "invoke"?

File: build.gradle [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+dependencies {
+    compile "io.netty:netty-all:${NETTY_ALL_VERSION}"
+    compile "ch.qos.logback:logback-classic:${LOGBACK_VERSION}"
+    compile "com.fasterxml.jackson.core:jackson-databind:${JACKSON_VERSION}"
+    compile "io.pravega:pravega-client:${PRAVEGA_VERSION}"
+}
+
+tasks.withType(JavaCompile) {
+    options.encoding = "UTF-8"
+}
```
[the remark this answers is not in the document]
Comment (vangork) [reply build.gradle]: done

File: gradle.properties [hunk 5]
```diff
@@ -0,0 +1,5 @@
+## dependencies
+NETTY_ALL_VERSION=4.1.51.Final
```
[the remark this answers is not in the document]
Comment (vangork) [reply gradle.properties]: done

File: src/main/java/io/pravega/ingest/gateway/IngestGateway.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        HttpServer server = new HttpServer(Parameters.getGatewayListeningPort());
+
+        // add router
+        server.router().post("/v1/scope/{scopeName}/stream/{streamName}/event", ApiController.class, "postEvent", AuthMiddleware.class);
+        server.router().post("/v1/scope/{scopeName}/stream/{streamName}/watermark", ApiController.class, "postWatermark", AuthMiddleware.class);
+
+        try {
+            server.start();
+        } catch (InterruptedException e) {
+            logger.error(e.getMessage());
```
[the remark this answers is not in the document]
Comment (vangork) [reply src/main/java/io/pravega/ingest/gateway/IngestGateway.java]: done

File: src/main/java/io/pravega/ingest/gateway/PravegaWriterManager.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.impl.ByteArraySerializer;
+import io.pravega.ingest.gateway.serializer.BytesSerializer;
+import java.net.URI;
+import java.util.HashMap;
+
+// TODO: add cache policy
+// https://blog.csdn.net/zane3/article/details/75563474
+public class PravegaWriterManager {
+    private static BytesSerializer SERIALIZER = new BytesSerializer();
```
[the remark this answers is not in the document]
Comment (vangork) [reply src/main/java/io/pravega/ingest/gateway/PravegaWriterManager.java]: done

File: src/main/java/io/pravega/ingest/gateway/PravegaWriterManager.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    private static HashMap<String, EventStreamWriter<byte[]>> cacheMap = new HashMap<String, EventStreamWriter<byte[]>>();
+
+    public static EventStreamWriter<byte[]> getOrCreate(String scope, String stream, String writerId) {
+        ByteArraySerializer SERIALIZER = new ByteArraySerializer();
```
[the remark this answers is not in the document]
Comment (vangork) [reply src/main/java/io/pravega/ingest/gateway/PravegaWriterManager.java]: done

File: src/main/java/io/pravega/ingest/gateway/server/http/HttpServer.java (hunk 1 above)
Comment (vangork) [reply src/main/java/io/pravega/ingest/gateway/server/http/HttpServer.java:37]: add comment for the params initialzation

File: src/main/java/io/pravega/ingest/gateway/server/http/HttpServerRequest.java (hunk 2 above)
Comment (vangork) [reply src/main/java/io/pravega/ingest/gateway/server/http/HttpServerRequest.java]: done

File: src/main/java/io/pravega/ingest/gateway/server/http/HttpServerRouter.java (hunk 3 above)
Comment (vangork) [reply src/main/java/io/pravega/ingest/gateway/server/http/HttpServerRouter.java]: you got me:)

**Comment by vangork** [comment]
> Additionally, there is a huge amount of HTTP server code. Was this hand written for this project or did it come from somewhere else? Are there base classes that we could inherit from or use directly? For instance, HttpServerRouter seems like something that any HTTP server library should provide.

All the classes for the project is hand written except for HTTPServer and HTTPServerInitialier take a reference from netty HTTP server sample. For the HTTP Server framework, I did some research and find several built on netty among which webflux framework seems tobe very popular, but it has a learning curve and i also doubt it might too heavy to have a complete framework for our usage.  So i decide to implement one by myself. I am following the basic concept of Lumen framework which a popular lightweight rest framework for PHP to design the class and interface. i am also thinking to support annotations for router definitions to make the code more simple and beautiful .

File: src/main/java/io/pravega/ingest/gateway/server/http/HttpServer.java (hunk 1 above)
Comment (claudiofahey) [reply src/main/java/io/pravega/ingest/gateway/server/http/HttpServer.java:37]: TCP_NODELAY=true does the opposite of what the comment explains.  You want this to be false which is the default. You should remove this entire line.

From the documentation:

> The default value of this socket option is FALSE. The socket option should only be enabled in cases where it is known that the coalescing impacts performance.

Comment (vangork) [reply src/main/java/io/pravega/ingest/gateway/server/http/HttpServer.java:37]: The configuration is correct, but i made a wrong comment. let mt modify the comment. we should disable Nagle's algorithm in low delay senario.

**Pull request merged**